### ISSUE TITLE
unable to stream or open USB device

### ISSUE THREAD

**Issue by raymanrd-cloud** [pschatzmann/arduino-audio-tools#2396]
### Problem Description

Hi - 
attempting to use just the usb-rx example -
it appears as a usb audio device - i can set the name etc - but i cannot actually use it as an audio device - it appears a s volume fixed at 0 / muted and when connecting to it from a player windows just wont - see image.

 i have successfully used and streamed audio wit hthe teensy libraries to teensy 4.0 on the same machine...

### Device Description

lilygo T-Display S3

### Sketch

```cpp
usb-rx example
```

### Other Steps to Reproduce

<img width="1925" height="885" alt="Image" src="https://github.com/user-attachments/assets/1a588a75-c4d7-4d40-a4ab-d8b60379c844" />

### What is your development environment (incl. core version info)

Arduino ESP32 3.3.11

### I have checked existing issues, discussions and online documentation (incl. the Wiki)

- [x] I confirm I have checked existing issues, discussions and online documentation

**Comment by raymanrd-cloud** [comment]
tried same code on the RP2040  - exactly the same code  - different target - 
it seems to work with the  RP, volume control comes up as adjustable and  able to play data to it at least  without  windows having a dummy spit ...
was also able  to get the parameters and set multi sample rate - although only 16 bit and only up to 48Khz.

It would seem that the issue is with the S3 version and in the initial usb negotiation - 
I was having the issue where  the S3 version would come up as a usb storage device regardlless of what I set - unless  i  did the initialization of the Streamcopy  with the source and targets  in the declaration static constructor - doing it as part of the begin didn't work - so looks like something is not right  in the early negotiation of the usb  ....

Hopefully some help
Rayman


note tried it with S2 as well - same as the S3 - so its an ESP32 issue -

**Comment by pschatzmann** [comment]
I havent used Windows for more then 20 year and I can confirm that the functionality is wokring pefeclty in Linux and OS X.
Since this is working on an RP2040, my guess would be that the bug must be somewhere in the TinyUSB layer provided by ESP32..

Just to double check: did you use the latst main branch ?

**Comment by raymanrd-cloud** [comment]
used the latest available form the dowload zip - in the code button - 

will see if theres any update to the the TinyUsb ..

**Comment by raymanrd-cloud** [comment]
some progress - but not there yet - 
did a whole round of library updates and... maybe windows decided to play somewhat - ?
now have the device available - can see it in windows,  at least via VLC - windows media player still doesnt like it apparently ....
in the control panel have volume slider and sample rates  - 

but 
can "send"" audio to it - eg from VLC - 
im using 
#define I2S_BCK_PIN   10
#define I2S_WS_PIN   11
#define I2S_DATA_PIN  12 and tried 13 also
on the usb_rx sketch - no mods, esp32s3 dev board
I see bclk on 10, 
I see lr_clk / WS on 11 correctly 
I see no DATA information on 12 or 13, just zero volt output - contiuuous . 

going back to square one and starting again .. . 

Cheers
Ray

**Comment by raymanrd-cloud** [comment]
oops realised im not quite using the exact emaple but a minor change 

/**
 * @file usb-rx.ino (speaker)
 * @brief USB Audio RX + USB CDC Serial on ESP32-S3 (composite device).
 *
 * The device appears to the host as a USB speaker (UAC2 RX — host sends
 * audio to the device). Audio throughput is measured and printed Serial via
 * MeasuringStream.
 *
 * @note for ESP32: Board settings (Arduino IDE):
 *   Tools → Board            : ESP32S3 Dev Module (or your board)
 *   Tools → USB Mode         : USB-OTG (TinyUSB)
 *   Tools → USB CDC On Boot  : Disabled
 *   If you want to use the CDC serial port for logging: define USBCDC
 * USBSerial; and use USBSerial!
 * 
 * @note for boards using Adafruit TinyUSB core (e.g. RP2040, STM32, SAMD):
 * Board settings (Arduino IDE): USB Mode: USB OTG (TinyUSB)
 *
 * @author Phil Schatzmann
 * @copyright GPLv3
 */
#include "AudioTools.h"
#include "AudioTools/Communication/USB/USBAudioStream.h"

// CDC output is pltform dependent
USBCDC MySerial;


#define I2S_BCK_PIN   10
#define I2S_WS_PIN   11
#define I2S_DATA_PIN  12


AudioInfo info(44100, 2, 16);
USBAudioStream in;
I2SStream       i2s_sink;  
StreamCopy copier(i2s_sink, in, 80);


void setup() {
  // Manual begin() is required on core without built-in support e.g. mbed
  // rp2040
  if (!TinyUSBDevice.isInitialized()) {
    TinyUSBDevice.begin(0);
  }
  // Register the CDC interface first (gets interfaces 0-1).
  MySerial.begin(115200);
  AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Warning);

  // Start MeasuringStream so it knows the audio format.

  Serial.println("Configuring I2S Interface for PCM5102...");
  auto i2s_cfg = i2s_sink.defaultConfig(TX_MODE);
  i2s_cfg.copyFrom(info);
  i2s_cfg.pin_bck = I2S_BCK_PIN;
  i2s_cfg.pin_ws = I2S_WS_PIN;
  i2s_cfg.pin_data = I2S_DATA_PIN;

   if (!i2s_sink.begin(i2s_cfg)) {
    Serial.println("Error initializing I2S framework.");
   //while(1);
  }

  // Register USB audio in RX mode (host → device, i.e. USB speaker).
  // begin_usb defaults to false so USB.begin() below controls the start.
  auto config = in.defaultConfig(RX_MODE);
  config.copyFrom(info);
  in.begin(config);

  // If already enumerated, additional class driver begin() e.g msc, hid, midi
  // won't take effect until re-enumeration: on ESP32 you can alternatively call USB.begin()
  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }

  MySerial.println("USB audio RX + CDC started");
}

void loop() {
  copier.copy();  // read ep_out_ff → MeasuringStream
  static unsigned long last_ms = 0;
  if (millis() - last_ms > 1000) {
    last_ms = millis();
    ;
    // MySerial.print("active="); MySerial.print(in.isStreamingActiveRx());
    // MySerial.print(" rxXfer="); MySerial.print(in.getRxXferCount());
    // MySerial.print(" rxBytes="); MySerial.print(in.getRxTotalBytes());
    // MySerial.print(" bufRx="); MySerial.print(in.bufferRx().available());
    // MySerial.print(" avail="); MySerial.println(in.available());
  }
}

**Comment by raymanrd-cloud** [comment]
validated that output side works to the i2s - using one of h]\the generators . - able to see data as well as the clocks,
so it seems that while the USB " initializes and does some level of connection " it doesn't seem that data is propagating from the PC usb connection into the unit .... as nothing comes out pf the "copier", but when using a generator  the data flows though the copier.

as I cant seem to get serial debug to work theres not much else I can do 

tried it as an android connection -  same result - visible as wired headphones - but no data flows through....

what version of the eso32 idf  did you last test with - i can do a rollback and see if that works ...

Ray

**Comment by pschatzmann** [comment]
I committed a correction:
- can you test with the original example
- does the usb-tx work for for you in windows ?

ps. please note that this is only working when you plug in the cable to the USB-OTG port. The usb-tx is easier to test since no CDC is needed and you just need to do an Audio Setup (discovery) and recoring in Audacity

**Comment by raymanrd-cloud** [comment]
installed your update - 
the usb TX works - ( may have worked before also )
RX seems to be able to now accept input from  m,edia player - 
but I have no way of telling with the default Rx as to whether data goes anywhere as all boards have a single usb-c port - I would need to get the soldering iron out to add a serial port - which i may yet do ...


So  I did the minimal mod of RX to get i2S output 
and am now seeing data  and get a valid i2s stream - confirmed with pcm5102 and headphones.

this was  the modified usb-rx


#include "AudioTools.h"
#include "AudioTools/Communication/USB/USBAudioStream.h"

// CDC output is pltform dependent
#ifdef ESP32
USBCDC MySerial;
#else
auto& MySerial = Serial;
#endif

#define I2S_BCK_PIN   10
#define I2S_WS_PIN    11
#define I2S_DATA_PIN  12


AudioInfo info(44100, 2, 16);
USBAudioStream in;
//MeasuringStream out(500, &MySerial);
I2SStream   out;
StreamCopy copier(out, in, 80);


void setup() {
  // Manual begin() is required on core without built-in support e.g. mbed
  // rp2040
  if (!TinyUSBDevice.isInitialized()) {
    TinyUSBDevice.begin(0);
  }
  // Register the CDC interface first (gets interfaces 0-1).
  MySerial.begin(115200);
  AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Warning);

  // Start MeasuringStream so it knows the audio format.
  auto i2s_cfg = out.defaultConfig(TX_MODE);
  i2s_cfg.copyFrom(info);
  i2s_cfg.pin_bck = I2S_BCK_PIN;
  i2s_cfg.pin_ws = I2S_WS_PIN;
  i2s_cfg.pin_data = I2S_DATA_PIN;
  out.begin(i2s_cfg);

  // Register USB audio in RX mode (host → device, i.e. USB speaker).
  // begin_usb defaults to false so USB.begin() below controls the start.
  auto config = in.defaultConfig(RX_MODE);
  config.copyFrom(info);
  in.begin(config);

  // If already enumerated, additional class driver begin() e.g msc, hid, midi
  // won't take effect until re-enumeration: on ESP32 you can alternatively call USB.begin()
  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }

  MySerial.println("USB audio RX + CDC started");
}

void loop() {
  copier.copy();  // read ep_out_ff → MeasuringStream
  static unsigned long last_ms = 0;
  if (millis() - last_ms > 1000) {
    last_ms = millis();
    MySerial.print("active="); MySerial.print(in.isStreamingActiveRx());
    MySerial.print(" rxXfer="); MySerial.print(in.getRxXferCount());
    MySerial.print(" rxBytes="); MySerial.print(in.getRxTotalBytes());
    MySerial.print(" bufRx="); MySerial.print(in.bufferRx().available());
    MySerial.print(" avail="); MySerial.println(in.available());
  }
}


So it looks like that fixed it !!

**Comment by pschatzmann** [comment]
Is CDC also working? You should get a Serial output via USB

**Comment by raymanrd-cloud** [comment]
yes - serial on CDC is now working.
I have successfully run with the audio on 1 core and lcd updating etc on the other core .

I did note - the only change I could make to the  source config was to increase fifo to 32. 
any other change  such as any of these 
  // usb_cfg.volume_active = true;
  //usb_cfg.product = " S3 Dac";
  //usb_cfg.serial = "12345";
  //usb_cfg.enable_multi_sample_rate = true;
  //usb_cfg.enable_interrupt_ep = true;

results in the media player dummy spit  - so appears to not get negotiated fully ....

BUT AT LEAST ITs working at the basic level !!

Cheers
Happy to test anything - particularly interested in the higher sample / bit depth options 

Ray

**Comment by pschatzmann** [comment]
I committed a correcton should make the enable_multi_sample_rate = true work also for ESP32

**Comment by raymanrd-cloud** [comment]
multi sample rates show and working -
 only goes up to 48K @ 16bit  so only 7 rates available of the potential 14..

Ray
